app/forms/routing/simulation_form.rb

```diff
@@ -81,32 +89,21 @@ def dst_country
 
   attr_reader :notices
 
-  def initialize(attrs = {})
-    @attrs = attrs
-    if attrs.present?
-      attrs.each do |k, v|
-        send("#{k}=", v) if respond_to?("#{k}=")
-      end
-    end
+  def debug
+    @debug&.map { |d| Result.new(d) }
   end
 
-  def auth_id=(val)
-    @auth_id = val.presence
-  end
+  protected
 
-  def has_attributes?
-    @attrs.present? && @attrs.to_unsafe_h.any?
+  def ip_is_valid
+    _tmp = IPAddr.new(remote_ip)
+  rescue IPAddr::Error => _error
+    errors.add(:remote_ip, 'is not valid')
   end
```

Comment:
create validator for ip_address

`app/validators/ip_address_validator.rb`

```
class IpAddressValidator < ActiveModel::EachValidator
 # ...
end
```

and use it like

```ruby
validate :remote_ip, ip_address: true
```

look at example at https://github.com/rails/rails/blob/2e62af310d98fba78f338d916f7c17996446bc40/activemodel/lib/active_model/validator.rb#L69